### PR TITLE
[AIRFLOW-3998] Use nested commands in cli.

### PR DESCRIPTION
**Pull request by jmcarp**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-3998
  - [AIP-16](https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-16%3A+Use+nested+commands+in+CLI)).

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

This is a proof of concept for https://issues.apache.org/jira/browse/AIRFLOW-3998 / https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-16%3A+Use+nested+commands+in+CLI.

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.
  - All the public functions and the classes in the PR contain docstrings that explain what it does

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

**Comment by zhongjiajie** [comment]
:+1: like to see this change.

File: airflow/bin/cli.py [hunk 1]
```diff
@@ -328,34 +348,44 @@ def pool_export_helper(filepath):
...
+def variables_delete(args):
+    Variable.delete(args.key)
+
+
+@cli_utils.action_logging
+def variables_import(args):
+    if os.path.exists(args.file):
+        import_helper(args.file)
+    else:
+        print("Missing variables file.")
```
Comment (mik-laj) [inline airflow/bin/cli.py:378]: What error code will be returned here?

**Comment by jmcarp** [comment]
This is still incomplete but ready for more review. I grouped commands related to dags into the `airflow dags` subcommand, tasks into `airflow tasks`, users into `airflow users`, etc. I don't have strong preferences about how we group some of these commands: for example, I left `airflow initdb` and `airflow resetdb` as top-level commands, but would be happy to move them to `airflow db init` and airflow db reset`.

@ashb @feng-tao @XD-DENG @Fokko: what do you think about the proposed changes? Happy to make any revisions, and I'll update documentation if/when we decide the final command structure.

File: airflow/bin/cli.py (hunk 1 above)
Comment (jmcarp) [reply airflow/bin/cli.py:378]: So far I haven't been changing the behavior of these commands, just changing the structure. Some error cases print a message and return without raising an error, and I agree that we should change that behavior. Any preference on whether that change happens as part of this patch or in a separate patch?

**Comment by ashb** [comment]
No strong preference to same commit/PR or others but maybe a slight lean towards another PR as this is already quite big.

Have you checked what the rendered CLI docs look like with this?

File: airflow/bin/cli.py [hunk 2]
```diff
@@ -258,38 +259,57 @@ def delete_dag(args):
...
-            else:
-                print("Missing pools file.")
-                pools = api_client.get_pools()
-        elif args.export:
-            pools = pool_export_helper(args.export)
-        else:
-            pools = api_client.get_pools()
-    except (AirflowException, IOError) as err:
-        log.error(err)
+@cli_utils.action_logging
```
Comment (ashb) [inline airflow/bin/cli.py]: list/get etc don't need this decorator, only things that modify state should have it (this wasn't possible before)

File: airflow/bin/cli.py [hunk 3]
```diff
@@ -330,34 +350,44 @@ def pool_export_helper(filepath):
...
+def variables_import(args):
+    if os.path.exists(args.file):
+        import_helper(args.file)
+    else:
+        print("Missing variables file.")
+
+
+@cli_utils.action_logging
+def variables_export(args):
+    export_helper(args.file)
```
Comment (ashb) [inline airflow/bin/cli.py]:
```suggestion
    variable_export_helper(args.file)
```

(and whatever is needed to make that function exist too)

File: airflow/bin/cli.py [hunk 4]
```diff
@@ -2186,73 +2140,217 @@ class CLIFactory(object):
...
-            'args': ('dag_id', 'subdir'),
-        }, {
-            'func': unpause,
-            'help': "Resume a paused DAG",
-            'args': ('dag_id', 'subdir'),
-        }, {
-            'func': trigger_dag,
-            'help': "Trigger a DAG run",
-            'args': ('dag_id', 'subdir', 'run_id', 'conf', 'exec_date'),
+            'help': 'DAGs',
```
Comment (ashb) [inline airflow/bin/cli.py]: Could we expand this help text a little bit? Even something like "List and manage DAGs" would be better

**Review by ashb — CHANGES_REQUESTED** [review]
Largely good I think, I only skimmed over the code changes (assuming they are some-what mechanical/rote in nature).

The following tasks are still at a top level:

- `task_failed_deps` -> should be under `tasks` group?
- `initdb`/`upgradedb`/`resetdb` - maybe a `db` group is called for?
- `sync_perm` -- it cuts across users, roles and dags, so this one probably should stay at the top level?
- `run` -- this somehow exists at the top level and under tasks.

Just running `airflow tasks` etc or any other "group" command fails with :`AttributeError: 'Namespace' object has no attribute 'func'` - that should print the help for the group ideally.

File: airflow/bin/cli.py (hunk 3 above)
Comment (jmcarp) [reply airflow/bin/cli.py]: Made this change.

File: airflow/bin/cli.py (hunk 2 above)
Comment (jmcarp) [reply airflow/bin/cli.py]: Thanks, I dropped the decorator from list/get functions that I added here. I see it's being used for other functions that don't appear to change state--I can clean those up as well or save for another patch.

File: airflow/bin/cli.py (hunk 4 above)
Comment (jmcarp) [reply airflow/bin/cli.py]: Updated.

**Comment by jmcarp** [comment]
@ashb: looks like sphinx-argparse handles nested commands correctly, and I updated so that missing subcommands print usage instructions instead of throwing an error. I also shuffled commands as suggested.

**Review by ashb — APPROVED** [review]

**Comment by jmcarp** [comment]
Updated documentation to match the new CLI structure.

What's the next step here? A vote on the AIP?

**Comment by ashb** [comment]
@jmcarp Yeah sounds right. Do you have time to write the vote email summarizing the changes or would you like me to pick this up?

**Comment by feng-tao** [comment]
the change LGTM, once we get the pass from AIP. I think it is good to go.

**Comment by jmcarp** [comment]
I'm traveling for work this week, so if you have time, feel free to write the vote email @ashb. Otherwise, I'll pick this up again next week.

**Comment by jmcarp** [comment]
@ashb: now that I have time to get back to this, I'm not sure where to find the procedure for holding the vote. How does that work?

**Comment by ashb** [comment]
It's not written down anywhere yet as we haven't fully formalized it. Here's an example of a recent vote on adding Pylint: https://lists.apache.org/thread.html/f4940d36e98ded96a2473bb2ccdfa4cc648faa2c1334b2aa901c0bba@%3Cdev.airflow.apache.org%3E

**Comment by ashb** [comment]
Started vote https://lists.apache.org/thread.html/cb2b854389ddc36d46e55d51625a90a54ef8d438299c5342bbd135bf@%3Cdev.airflow.apache.org%3E

**Comment by jmcarp** [comment]
Updated, tests passing.

**Comment by jmcarp** [comment]
Conflicts resolved. Ready for review @ashb @mik-laj or anybody else interested.

**Comment by ashb** [comment]
Sorry @jmcarp, I was in inbox hell and didn't see the ping, and it's conflicted again. I'm happy with this change now and the AIP is passed so once you resolve the conflicts again you can merge yourself(?)

**Comment by jmcarp** [comment]
Conflicts resolved @ashb. I don't think I ever got commit access to the repo--I must have missed a step. Any idea what I missed?

**Comment by mik-laj** [comment]
@jmcarp https://gitbox.apache.org/setup/ ;-)

**Comment by mik-laj** [comment]
I'm waiting for Travis to finish work.

**Pull request merged**